### PR TITLE
Make MateImageMenuItem subclassable

### PR DESCRIPTION
**Pull request by zhuyaliang**
See the discussion on:https://github.com/mate-desktop/mate-panel/issues/1433

### CODE REVIEW

**Review by lukefromdc — COMMENTED** [review]
I tested a build of this, nothing broke but when I tried to replace GtkImageMenuItem with MateImageMenuItem in mate-panel's panel-menu-items.h I still got 

```
panel-menu-items.h:46:9: error: unknown type name ‘MateImageMenuItem’
   46 |         MateImageMenuItem            menuitem;
      |         ^~~~~~~~~~~~~~~~~
panel-menu-items.h:52:9: error: unknown type name ‘MateImageMenuItemClass’
   52 |         MateImageMenuItemClass       menuitem_class;
      |         ^~~~~~~~~~~~~~~~~~~~~~
panel-menu-items.h:69:9: error: unknown type name ‘MateImageMenuItem’
   69 |         MateImageMenuItem            menuitem;
      |         ^~~~~~~~~~~~~~~~~
panel-menu-items.h:75:9: error: unknown type name ‘MateImageMenuItemClass’
   75 |         MateImageMenuItemClass       menuitem_class;
      |         ^~~~~~~~~~~~~~~~~~~~~~

```
Note that I do not really understand how subclassing works so I don't know if the problem was on my end or here.

Code giving these errors was

```
struct _PanelPlaceMenuItem {
	MateImageMenuItem            menuitem;

	PanelPlaceMenuItemPrivate  *priv;
};

struct _PanelPlaceMenuItemClass {
	MateImageMenuItemClass       menuitem_class;
};

```
and

```
struct _PanelDesktopMenuItem{
	MateImageMenuItem            menuitem;

	PanelDesktopMenuItemPrivate  *priv;
};

struct _PanelDesktopMenuItemClass {
	MateImageMenuItemClass       menuitem_class;
};

```

**Comment by zhuyaliang** [comment]
@lukefromdc Need to add header file in ```panel-menu-items.h```
```
#include <libmate-desktop/mate-image-menu-item.h>
```

**Comment by zhuyaliang** [comment]
```
diff --git a/mate-panel/panel-menu-items.c b/mate-panel/panel-menu-items.c
index 14b58701..ba61a4b2 100644
--- a/mate-panel/panel-menu-items.c
+++ b/mate-panel/panel-menu-items.c
@@ -41,7 +41,6 @@
 #include <glib/gi18n.h>
 #include <gio/gio.h>
 #include <libmate-desktop/mate-gsettings.h>
-#include <libmate-desktop/mate-image-menu-item.h>
 
 #include <libpanel-util/panel-error.h>
 #include <libpanel-util/panel-glib.h>
diff --git a/mate-panel/panel-menu-items.h b/mate-panel/panel-menu-items.h
index 1b94a2d8..90412ef8 100644
--- a/mate-panel/panel-menu-items.h
+++ b/mate-panel/panel-menu-items.h
@@ -26,6 +26,7 @@
 
 #include <gtk/gtk.h>
 #include "panel-widget.h"
+#include <libmate-desktop/mate-image-menu-item.h>
 
 #ifdef __cplusplus
 extern "C" {
```

**Comment by lukefromdc** [comment]
Thanks, just plain forgot that include!

**Review by lukefromdc — APPROVED** [review]
Builds and runs fine, and using MateImageMenuItem in mate-panel/mate-panel/panel-menu-items.h to finish the other two commits porting from GtkImageMenuItem to MateImageMenuItem now works. The panel build finished clean, and dragging an icon from the compact menu to the panel doesn't crash the panel

**Comment by lukefromdc** [comment]
This seems to be the correct way to fix this mess. Question is: do we want to release mate-desktop 1.28.2 and a version of mate-panel depending on that (a breaking change within a major version release), do we want to bump versions to 1.29 for this and use the fix I have working (it's a clunky workaround but functions) for the brand-new 1.28 release, or something else? That question is above my "pay grade."

I don't know if any distros have even picked up 1.28 yet, but if they have not, we could get this out the door in such a way that no distro would have to package a version of mate-panel depending on a mate-desktop version newer than what they first put out for 1.28

File: libmate-desktop/mate-image-menu-item.c [hunk 1]
```diff
@@ -22,27 +22,27 @@
 
 #include "mate-image-menu-item.h"
 
-struct _MateImageMenuItem
+struct _MateImageMenuItemPrivate
 {
-    GtkMenuItem  menu_item;
-    GtkWidget   *image;
+  GtkWidget *image;
```
Comment (cwendling) [inline libmate-desktop/mate-image-menu-item.c:27]: minor nitpick: indentation should be 4 spaces, not 2

File: libmate-desktop/mate-image-menu-item.h [hunk 2]
```diff
@@ -23,8 +23,35 @@
...
+typedef struct _MateImageMenuItem              MateImageMenuItem;
+typedef struct _MateImageMenuItemPrivate       MateImageMenuItemPrivate;
+typedef struct _MateImageMenuItemClass         MateImageMenuItemClass;
+
+struct _MateImageMenuItem
+{
+  GtkMenuItem menu_item;
+
+  /*< private >*/
+  MateImageMenuItemPrivate *priv;
```
Comment (cwendling) [inline libmate-desktop/mate-image-menu-item.h:41]: same nitpick for the indentation

File: libmate-desktop/mate-image-menu-item.h [hunk 3]
```diff
@@ -23,8 +23,35 @@
...
+};
+
+/**
+ * MateImageMenuItemClass:
+ * @parent_class: The parent class.
+ */
+struct _MateImageMenuItemClass
+{
+  GtkMenuItemClass parent_class;
+
```
Comment (cwendling) [inline libmate-desktop/mate-image-menu-item.h:51]: Here as well, plus one unnecessary blank line

**Review by cwendling — APPROVED** [review]
Apart from the indentation nitpick, this looks great and seem to work fine, well done! :+1:

**Comment by cwendling** [comment]
> This seems to be the correct way to fix this mess. Question is: do we want to release mate-desktop 1.28.2 and a version of mate-panel depending on that (a breaking change within a major version release), do we want to bump versions to 1.29 for this and use the fix I have working [mate-desktop/mate-panel#1434](https://github.com/mate-desktop/mate-panel/pull/1434) (it's a clunky workaround but functions) for the brand-new 1.28 release, or something else? That question is above my "pay grade."

Well, if it was only the revert, I would have said just revert in 1.28 and leave this fix for 1.29.
But if it requires additional changes to display the icons as you added in mate-panel#1434, I think it's wiser to include all this in a 1.28.x and adjust the dependency.

The changes here are actually fairly trivial, and the changes in mate-panel to use this would as well (as you talked about above), and I actually feel a bit more comfortable with that than the extra copy and code in mate-panel#1434.

So my vote goes for integrating this in a 1.28 point release.

> I don't know if any distros have even picked up 1.28 yet, but if they have not, we could get this out the door in such a way that no distro would have to package a version of mate-panel depending on a mate-desktop version newer than what they first put out for 1.28

Whether or not anybody packaged it is fairly irrelevant IMO, and we should not try and take shortcuts in that might or might not help anybody, but would likely cause issues for others.
What we need to do is do something coherent that don't allow for incompatible versions to be used together, that is release a new version of the affected component, and update the dependencies accordingly.  So yes, distros would have to package the new mate-desktop first, and then build the new mate-panel with it, but that's nothing out of the ordinary for any build.

**Comment by raveit65** [comment]
Let us do simply do a new 1.28 point release when this is merged.
I would expect that only fedora 40 (which is in before beta state) and maybe archlinux (rolling release) use the broken 1.28.0. And i've already reverted the 2 bad commits for fedora since 2 weeks :-)
So updating the unresolved conversation and i will do a rush new  point release.
No need to make it more complicated as needed.

**Comment by raveit65** [comment]
We can add to release notes and News  that a rebuild of mate-panel is needed.
I am against to bumping up requires in this early release state as it makes it hard to downgrade a package.

**Review by raveit65 — APPROVED** [review]
LGTM

**Comment by lukefromdc** [comment]
Remember that the previous mate-panel commit stopped the crashes by making MateImageMenuItem the same size as GtkImageMenuItem, but we still need to apply a commit to mate-panel to use it panel-menu-items.h to actually use MateImageMenuItem properly and not get another crash if a future GTK version makes a change in GtkImageMenuItem or simply breaks it

**Comment by cwendling** [comment]
> I can confirm that this PR and a rebuild of mate-panel fixes issue [mate-desktop/mate-panel#1433](https://github.com/mate-desktop/mate-panel/issues/1433) and `mate-panel --replace` command. So after format fixes this is ready to go.

With only this PR (e.g. not changing mate-panel to use the right struct type), note that the fact it doesn't crash anymore is again *by chance* because `MateImageMenuItem` now so happens to have the same size as `GtkImageMenuItem`.

> We can add to release notes and News that a rebuild of mate-panel is needed. I am against to bumping up requires in this early release state as it makes it hard to downgrade a package.

This is IMO a bad idea, because it means the dependencies are not correct.  What's the issue of having proper dependencies?  Yes, we could get away with it because it just so happens that updating mate-panel is "optional" by sheer luck, but the conceptually correct fix is using the right type in mate-panel, meaning it depends on a mate-desktop exposing it.

**Comment by lukefromdc** [comment]
If we have a new version of mate-panel as well, with the added commit to panel-menu-items.h to actually use MateImageMenuItem and depending on mate-desktop 1.28.2 we will have the correct dependencies. Will make a mate-panel PR now

**Comment by lukefromdc** [comment]
https://github.com/mate-desktop/mate-panel/pull/1437 is a companion to this

**Comment by raveit65** [comment]
Ok, do what you want . I am leaving Mate until now.
Bye bye, and thank you for the fish

**Comment by lukefromdc** [comment]
I am also OK with moving this to a 1.29 branch and reopening https://github.com/mate-desktop/mate-panel/pull/1434 but since I have so little interaction with the distro teams outside discussions here I don't feel I am the one to make the decision. I am OK with either way of doing this, what is basically an accident of timing set this up.

**Comment by cwendling** [comment]
> Ok, do what you want . I am leaving Mate until now. Bye bye, and thank you for the fish

@raveit65 Not sure how serious you are, but anyway:  I don't want to upset anybody, and I never prevented anything from being merged I believe. I'm just giving my opinion, and explaining why I think what I do.  Usually, if I'm concerned enough, I contribute the code as soon as life allows me to do what I think is right, but I don't force anything either.
I'm sorry if this is going through another way, but I'm not judging anybody here -- and admittedly I am leaning too far towards "conceptually correct" for my own good.

In any case, I won't fight, and won't bother anybody further on this topic: I gave my opinion and explained why, and if you think it's better to do otherwise, please do.

---

> I am also OK with moving the mate-panel change to a 1.29 branch and reopening https://github.com/mate-desktop/mate-panel/pull/1434 but since I have so little interaction with the distro teams outside discussions here I don't feel I am the one to make the decision.
>
> […]
>
> Therefore, we can merge this one, bump version, and release without breaking anything no matter what we decide to do with the panel.

@lukefromdc Yes, in **practice**, we could simply publish this PR in a 1.28 point release and it would magically fix mate-panel (I think) because the struct sizes would match and nothing would access invalid memory anymore.  So if upgrading inter-dependencies inside a stable release is a problem, it's an option.  So we don't really need https://github.com/mate-desktop/mate-panel/pull/1434 either way.

Note that we then need *not* include https://github.com/mate-desktop/mate-panel/pull/1437 in 1.28, as that would depend on this PR.

> If you all are OK with it, I can fix the format issues myself and merge this

Sure on my end, and I would *think* that this part is uncontroversial.

**Comment by raveit65** [comment]
>  Not sure how serious you are, but anyway: I don't want to upset anybody, and I never prevented anything from being merged I believe. I'm just giving my opinion, and explaining why I think what I do. Usually, if I'm concerned enough, I contribute the code as soon as life allows me to do what I think is right, but I don't force anything either.
I'm sorry if this is going through another way, but I'm not judging anybody here -- and admittedly I am leaning too far towards "conceptually correct" for my own good.

>In any case, I won't fight, and won't bother anybody further on this topic: I gave my opinion and explained why, and if you think it's better to do otherwise, please do.

Forget all that nonsense what do you think about me.
I was the first one who provide a public repository of Mate for linux in 2014.
Sadly i did to much work for Mate in the last 5 years, to much releases which i did alone without any help. To much stupid discussions with users. Since half a year my plan was to get out of this trap and releases 1.28 and say good bye.
And 1.28 is done for me. All that minor point releases you can do for yourself. All tools are in your hand ...... and you both are in owner group of MATE at github
This decision has nothing to do with this report but i am really tired of this problems :-)
Don't forget to maintain Mate website (owner is Martin Wimpress) , transifex, travis Ci and everything else.
Until now my role is downstream maintainer only. And it is possible that i will remove Mate from fedora 42 when 1.28 is too old.
Of course I will disable github notifications to avoid all that spam in my mailbox. If you have questions you can send me an email.
Good Luck

**Comment by zhuyaliang** [comment]
@raveit65 You have done a lot of work for ```Mate Deskto````p, and ```Mate Desktop``` needs you. Please do not leave this team. We have been working hard to provide users with services for many years, so please do not give up

**Comment by cwendling** [comment]
@raveit65 I hear you that maintenance burden can grow heavy, especially if you feel some tasks are too heavily relying on you although you asked for others to help with it -- and you did.  I hope you'll reconsider at some point, but anyway thanks a lot for all the work you've done :)

**Review by cwendling — APPROVED** [review]
LGTM

**Pull request merged**